packages/wrangler/src/user/auth-variables.ts

```diff
@@ -0,0 +1,151 @@
+ *
+ * Normally you should not need to set this explicitly.
+ * If you want to switch to the staging environment set the
+ * `WRANGLER_USE_STAGING` environment variable instead.
+ */
+export const getTokenUrlFromEnv = getEnvironmentVariableFactory({
+	variableName: "WRANGLER_TOKEN_URL",
+	defaultValue:
+		getCloudflareApiEnvironmentFromEnv() === "staging"
+			? "https://dash.staging.cloudflare.com/oauth2/token"
+			: "https://dash.cloudflare.com/oauth2/token",
+});
+
+/**
+ * `WRANGLER_REVOKE_URL` is the path that is used to exchange an OAuth
+ * refresh token for a new OAuth token.
+ *
+ * Normally you should not need to set this explicitly.
+ * If you want to switch to the staging environment set the
+ * `WRANGLER_USE_STAGING` environment variable instead.
+ */
+export const getRevokeUrlFromEnv = getEnvironmentVariableFactory({
+	variableName: "WRANGLER_REVOKE_URL",
+	defaultValue:
+		getCloudflareApiEnvironmentFromEnv() === "staging"
+			? "https://dash.staging.cloudflare.com/oauth2/revoke"
+			: "https://dash.cloudflare.com/oauth2/revoke",
+});
+
+/**
+ * Set the `WRANGLER_CF_AUTHORIZATION_TOKEN` to the CF_Authorization token found at https://dash.staging.cloudflare.com/bypass-limits
+ * if you want to access the staging environment, triggered by `WRANGLER_API_ENVIRONMENT=staging`.
+ */
+export const getCloudflareAccessToken = () => {
+	const env = getEnvironmentVariableFactory({
+		variableName: "WRANGLER_CF_AUTHORIZATION_TOKEN",
+	})();
+
+	// If the environment variable is defined, go ahead and use it.
+	if (env !== undefined) {
+		return env;
+	}
+
+	const cloudflareAuthHost = new URL(getTokenUrlFromEnv()).host;
+
+	// Try to get the access token via cloudflared
+	try {
+		const { stdout: token } = execaSync("cloudflared", [
+			`access`,
+			`token`,
+			`--app`,
+			cloudflareAuthHost,
+		]);
+		if (
+			!token.includes(
+				"Unable to find token for provided application. Please run login command to generate token."
+			)
+		) {
+			return token;
+		}
+	} catch (e) {
+		// OK that didn't work... move on.
+		logger.debug(e);
+	}
+
+	// No luck. Let's try to get it by logging in via cloudflared.
+	try {
+		const { stdout: login } = execaSync("cloudflared", [
+			`access`,
+			`login`,
+			cloudflareAuthHost,
+		]);
+		const match = /Successfully fetched your token:\s+(.+)/.exec(login);
+		if (match) {
+			console.log(match);
+			return match[1];
+		}
+	} catch (e) {
+		// Didn't work either... moving along.
+		logger.debug(e);
+	}
+	// Still no luck give up and give the user some ideas of next steps.
+	throw Error(
+		"When trying to access staging environment we need an 'access token'.\n" +
+			"We were unable to get one automatically using cloudflared. Run with debug logging to see more detailed errors.\n" +
+			"Alternatively, you could provide your own access token by setting the WRANGLER_CF_AUTHORIZATION_TOKEN environment variable\n" +
+			"to a token that is generated at https://dash.staging.cloudflare.com/bypass-limits."
+	);
+};
```

Comment:
It was landed after this PR was opened (sorry!), but could you use the helper function `getAccessToken` in `packages/wrangler/src/user/access.ts` for this logic?

Reply:
Will do. Thanks for noticing.

Reply:
Ugh... `getAccessToken()` is async, but these functions are sync...

Reply:
Oh no, it's OK.